### PR TITLE
Add user note when recurring payment fails. #990

### PR DESCRIPTION
**Pull request by nosegraze**
#990

### CODE REVIEW

File: includes/member-functions.php [hunk 1]
```diff
@@ -1284,4 +1284,24 @@ function rcp_update_expired_member_role( $status, $member_id, $old_status, $memb
...
+ *
+ * @param RCP_Member          $member
+ * @param RCP_Payment_Gateway $gateway
+ *
+ * @since 2.7.4
+ * @return void
+ */
+function rcp_add_recurring_payment_failure_note( $member, $gateway ) {
+	if ( 'RCP_Payment_Gateway_Stripe' != get_class( $gateway ) && 'RCP_Payment_Gateway_Stripe_Checkout' != get_class( $gateway ) ) {
+		rcp_add_member_note( $member->ID, __( 'Recurring charge failed.', 'rcp' ) );
```
Comment (pippinsplugins) [inline includes/member-functions.php]: Could we move the note left for Stripe down here too? That way it's all in one place.

**Comment by nosegraze** [comment]
Looks like `$event` was added to this hook in another PR:

`do_action( 'rcp_recurring_payment_failed', $member, $this, $event );`

Will that cause issues if we don't pass `$event` in any other gateway? Or can we just pass it in some but not others?

**Comment by nosegraze** [comment]
Removed `$event` from `rcp_recurring_payment_failed`. It was added in https://github.com/restrictcontentpro/restrict-content-pro/pull/985 but I don't think it should be there in the non-Stripe-specific actions. It's still in `rcp_stripe_charge_failed` and can be used there.

**Comment by tnorthcutt** [comment]
@nosegraze Since `$event` was only added in `class-rcp-payment-gateway-stripe.php`, it would only fire if Stripe is the acting gateway, right? Any downside to having it in there?

**Comment by nosegraze** [comment]
Well `rcp_recurring_payment_failed` exists in all gateways. So I guess my worry was with mismatching parameter numbers if someone hooks into that for more than one gateway and it's expecting 2 parameters sometimes and 3 other times. I feel like I've seen warnings about that before but I'm not really sure.

Given that there's another action in there that _does_ include `$event` I think I'd rather omit it from `rcp_recurring_payment_failed` for consistency across all gateways.

**Comment by mindctrl** [comment]
@nosegraze @pippinsplugins some questions and thoughts.

Why would Stripe have a standalone and not the others? As time goes on, we'll probably want to add the event ID in the note from other gateways, which would mean making a separate function for each if we follow this structure. Which means multiple functions fire and check for the proper gateway in every one of them, like the first line does in `rcp_add_recurring_payment_failure_note()`. Which gives us a bunch of gateway check code duplication just to add a payment failed note.

I'd love to see one function adding the note no matter the gateway used.

One idea is to save the relevant event ID, txn ID, whatever the gateway calls it, from the webhook as a `$this->webhook_event_id` variable before passing the gateway object to the hook. That would let one function add the note and reference the normalized variable name inside the object to get the event ID.

**Comment by nosegraze** [comment]
> One idea is to save the relevant event ID, txn ID, whatever the gateway calls it, from the webhook as a $this->webhook_event_id variable before passing the gateway object to the hook.

I like that idea.

**Comment by mindctrl** [comment]
Updated the PR with the idea mentioned above. Need to test the various gateways to make sure the proper transaction ID is being used.

**Comment by nosegraze** [comment]
Looks good! Just added one small addition.

**Comment by pippinsplugins** [comment]
I really like the way this has evolved :+1:

File: includes/member-functions.php [hunk 2]
```diff
@@ -1312,4 +1312,32 @@ function rcp_update_expired_member_role( $status, $member_id, $old_status, $memb
...
 	}
 }
-add_action( 'rcp_set_status', 'rcp_update_expired_member_role', 10, 4 );
\ No newline at end of file
+add_action( 'rcp_set_status', 'rcp_update_expired_member_role', 10, 4 );
+
+/**
+ * Add a note to the member when a recurring charge fails.
+ *
+ * Stripe and Stripe Checkout are handled separately to include the event ID.
```
Comment (pippinsplugins) [inline includes/member-functions.php]: This comment needs removed since it has been standardized.

**Pull request merged**